modules/internal/ChapelDistribution.chpl

```diff
@@ -657,4 +771,48 @@ module ChapelDistribution {
       }
     }
   }
+
+  // delete helpers
+
+  // param privatized here is a workaround for the fact that
+  // we can't include the privatized freeing for DefaultRectangular
+  // because of resolution order issues
+  proc _delete_dist(dist:BaseDist, param privatized:bool) {
+    //extern proc printf(fmt:c_string);
+    //printf("in _delete_dist privatized=" + privatized:string + "\n");
+    dist.dsiDestroyDist();
+
+    if privatized {
+      _freePrivatizedClass(dist.pid, dist);
+    }
+
+    delete dist;
+  }
+
+  proc _delete_dom(dom:BaseDom, param privatized:bool) {
+    dom.dsiDestroyDom();
+
+    if privatized {
+      _freePrivatizedClass(dom.pid, dom);
+    }
+
+    delete dom;
+  }
+  // arr is a subclass of :BaseArr but is generic so
+  // that arr.eltType is meaningful.
+  proc _delete_arr(arr, param privatized:bool) {
+    // unlink domain referred to by arr.eltType
+    chpl_decRefCountsForDomainsInArrayEltTypes(arr, arr.eltType);
+    // decide whether or not the array is an alias
+    var isalias = (arr._arrAlias != nil);
+    // array implementation can destroy data or other members
+    arr.dsiDestroyArr(isalias);
```

Comment:
dsiDestroyArr is overridden by various array implementations. E.g. DefaultRectangular, Block, etc. I thought it better to include an argument than require these to check _arrAlias because I think that element of the array destruction story is more in the domain of the base implementation than the derived ones.

Nonetheless AFAIK only DefaultRectangular uses the argument now.